util/patricia_trie/src/triedb.rs

```diff
@@ -39,45 +42,51 @@ use bytes::Bytes;
 /// use hashdb::*;
 /// use memorydb::*;
 /// use ethereum_types::H256;
+/// use node_codec::RlpNodeCodec;
+///
+/// type RlpCodec = RlpNodeCodec<KeccakHasher>;
 ///
 /// fn main() {
-///   let mut memdb = MemoryDB::new();
+///   let mut memdb = MemoryDB::<KeccakHasher>::new();
 ///   let mut root = H256::new();
-///   TrieDBMut::new(&mut memdb, &mut root).insert(b"foo", b"bar").unwrap();
-///   let t = TrieDB::new(&memdb, &root).unwrap();
+///   TrieDBMut::<_, RlpCodec>::new(&mut memdb, &mut root).insert(b"foo", b"bar").unwrap();
+///   let t = TrieDB::<_, RlpCodec>::new(&memdb, &root).unwrap();
 ///   assert!(t.contains(b"foo").unwrap());
 ///   assert_eq!(t.get(b"foo").unwrap().unwrap(), DBValue::from_slice(b"bar"));
 /// }
 /// ```
-pub struct TrieDB<'db> {
-	db: &'db HashDB,
-	root: &'db H256,
+pub struct TrieDB<'db, H, C>
+where 
+	H: Hasher + 'db, 
+	C: NodeCodec<H>
+{
+	db: &'db HashDB<H>,
+	root: &'db H::Out,
 	/// The number of hashes performed so far in operations on this trie.
 	hash_count: usize,
+	codec_marker: PhantomData<C>,
 }
 
-impl<'db> TrieDB<'db> {
+impl<'db, H, C> TrieDB<'db, H, C>
```

Comment:
you should add another `impl` to simplify init of this structure

```rust
impl<'db> TrieDB<'db, RlpCodec, KeccakHasher> {
  fn with_rlp_codec(db: &'db HashDB<KeccakHasher>, root: &db H256) -> super::Result<Self> {
    Self::new(db, root)
  }
}

```

Reply:
I disagree, that would happen naturally if we had a `type TrieDB<'db> = generic::TrieDB<'db, RlpCodec, KeccakHasher>` in the `patricia-trie-ethereum` crate I suggested

Reply:
I favor a split because we want to depend on patricia trie logic but not keccak or rlp in Polkadot.

Reply:
you're right. I read your review after I did mine :)